### DISCUSSION TITLE
efficient way to read complete stream?

### DISCUSSION THREAD

**Discussion by allefeld** [PyAV-Org/PyAV#858] [1. Help]
I'm trying to read a complete audio stream from a video file into a numpy array:

```python
container = av.open(videoFile)
audioStream = container.streams.audio[audioIndex]
audio = []
for frame in container.decode(audioStream):
    audio.append(frame.to_ndarray())
audio = np.hstack(audio)
```

Is this the correct way, or is it possible to do it in a more efficient way?
I thought about preallocating the numpy array, but `audioStream.frames` is 0 and `audioStream.frames` is `None`.

**Comment by shersonb** [comment]
If you are attempting to read and decode the entire stream, iterating over container.decode(audioStream) is the way to do it. However, is there a technical reason you must load the entire decoded audio stream into memory at once?

**Comment by allefeld** [comment]
I want to perform alignment (both shifting & scaling) of the audio stream with another time series based on cross-correlation. I think reading all data is the best way to do so, yes.

The reason for my question is that each frame only contains 512 samples (at 48kHZ approx. 11 ms), and the whole stream is 43 min long. I feel like having larger chunks of data may be better. But as far as I can tell, there is no way to change that, or to read several frames at once?

The other part is whether collecting subarrays in a list and then using `np.hstack` is the best way to collect the read data?
